### PR TITLE
push context: handle panic if there is an error in resource initialization

### PR DESCRIPTION
**Pull request by ramaraochavali**
There was an error in one of the Virtual Services in our env which was added by mistake to K8S and it caused Pilot to Panic - bringing down the whole cluster.  I think Pilot should handle crash while initializing/updating push context where it process all the external resources - they can be in bad state for variety of reasons.

### CODE REVIEW

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -729,6 +730,12 @@ func (ps *PushContext) InitContext(env *Environment, oldPushContext *PushContext
 }
 
 func (ps *PushContext) createNewContext(env *Environment) error {
+
+	// If any of the custom CRD initialization causes panic, we should recover.
```
Comment (ramaraochavali) [inline pilot/pkg/model/push_context.go]: Does it make sense to continue with push here?

File: pilot/pkg/model/push_context.go [hunk 2]
```diff
@@ -770,6 +777,11 @@ func (ps *PushContext) updateContext(
 	oldPushContext *PushContext,
 	pushReq *PushRequest) error {
 
+	// If any of the custom CRD initialization causes panic, we should recover.
+	defer goutils.HandleCrash(func() {
```
Comment (ramaraochavali) [inline pilot/pkg/model/push_context.go]: In update case, does it make sense to set the oldcontext back?

**Comment by howardjohn** [comment]
seems like we should just be ensuring the crd are correct in validation. Why is the push context special here, can a bad crd only crash here?

**Comment by ramaraochavali** [comment]
Agree on we should ensure crd are correct in validation. I guess we don’t
do schema validation- do we?
Push context is the major consumer of them - hence special care.

But definitely open for discussion - and I am happy to close this if this
is not the right approach.

**Comment by howardjohn** [comment]
You are not using galley, right? I think in previous versions we didn't do validation without galley, but someone fixed it. If thats not happening we should definitely fix ti.

I just worry here we get in a bad state and all pushes stop but we don't know why. At least if its crashlooping we will notice

**Comment by ramaraochavali** [comment]
Yes.  We do not use galley. Schema validation I
think is not happening- but individual field validation is happening. I
will a look at that part.
 Agree on the crash loop part - was thinking about same while making this
change.

**Comment by ramaraochavali** [comment]
I will close this - will look at validation part instead

**Pull request closed without merging**